server/src/main/java/org/apache/iotdb/db/metadata/storagegroup/StorageGroupManager.java

```diff
@@ -0,0 +1,241 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.metadata.storagegroup;
+
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+import org.apache.iotdb.db.exception.metadata.MetadataException;
+import org.apache.iotdb.db.exception.metadata.StorageGroupNotSetException;
+import org.apache.iotdb.db.metadata.MManager;
+import org.apache.iotdb.db.metadata.mnode.IStorageGroupMNode;
+import org.apache.iotdb.db.metadata.mtree.MTreeAboveSG;
+import org.apache.iotdb.db.metadata.path.PartialPath;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.iotdb.db.conf.IoTDBConstant.PATH_ROOT;
+import static org.apache.iotdb.db.conf.IoTDBConstant.PATH_SEPARATOR;
+
+// This class implements all the interfaces for storage group management. The MTreeAboveSg is used
+// to manage all the storage groups and MNodes above storage group.
+public class StorageGroupManager implements IStorageGroupManager {
+
+  private static final Logger logger = LoggerFactory.getLogger(StorageGroupManager.class);
+
+  private MTreeAboveSG mtree;
+
+  public StorageGroupManager() {
+    mtree = new MTreeAboveSG();
+  }
+
+  public synchronized void init() {
+    mtree.init();
+    File dir = new File(IoTDBDescriptor.getInstance().getConfig().getSchemaDir());
+    File[] sgDirs = dir.listFiles((dir1, name) -> name.startsWith(PATH_ROOT + PATH_SEPARATOR));
+    if (sgDirs != null) {
+      for (File sgDir : sgDirs) {
+        try {
+          setStorageGroup(new PartialPath(sgDir.getName()));
+        } catch (MetadataException e) {
+          logger.error("Cannot recover storage group from dir {} because", sgDir.getName(), e);
+        }
+      }
+    }
+  }
+
+  public synchronized void clear() {
+    for (SGMManager sgmManager : getAllSGMManagers()) {
+      sgmManager.clear();
+    }
+    if (mtree != null) {
+      mtree.clear();
+    }
+  }
+
+  @Override
+  public void setStorageGroup(PartialPath path) throws MetadataException {
+    IStorageGroupMNode storageGroupMNode = mtree.setStorageGroup(path);
+    SGMManager sgmManager = new SGMManager();
+    storageGroupMNode.setSGMManager(sgmManager);
+    sgmManager.init(storageGroupMNode);
+  }
+
+  @Override
+  public SGMManager getBelongedSGMManager(PartialPath path) throws MetadataException {
```

Comment:
Not sure whether this method is ok with lock-free. We need to ensure this method could be invoked in parallel.

The root node of SGMManager will be changed when this method is invoked, we need to consider whether it is ok here without concurrent control.

Reply:
The node change is just the nodeType change, from storageGroupMNode to storageGroupEntityMNode, the reference of SGMManager will be taken by both the old node and new node, thus there's no concurrent problem. Since all the references points to the old node will re-point to the new node, thus the old node will be collected as a garbage.